File: client/python/features/steps/rename.py
from asyncio import sleep
from addict import Dict
from behave import then, step
from behave.api.async_step import async_run_until_complete
from behave.runner import Context
from bson import ObjectId
from bson.json_util import dumps
from hamcrest import assert_that, equal_to, not_

from client import has_key as _has_key
from logger import log as _log
from request import create_request, drop_collection_request, rename_collection_request, retrieve_request

_database = "itest"
_collection = "oldname"
_target = "newname"


@then("Document created in a new collection")
@async_run_until_complete
async def step_impl(context: Context):
    assert_that(hasattr(context, "client"), "Client not set")
    context.oid = ObjectId()
    req = create_request(database=_database, collection=_collection, document=Dict({"_id": context.oid, "key": "value"}))
    resp = await context.client.execute(req)
    _log.info(f"Create response: {dumps(resp)}")

    assert_that(_has_key("error", resp), not_(True), "Response contained error")
    assert_that(_has_key("database", resp), "Response did not contain version history database")
    assert_that(_has_key("collection", resp), "Response did not contain version history collection")
    assert_that(_has_key("_id", resp), "Response did not contain version history id")
    assert_that(_has_key("entity", resp), "Response did not include created entity information")
    assert_that(resp.entity, equal_to(context.oid), "Response entity id not same as specified")

    context.vhdb = resp.database
    context.vhc = resp.collection
    context.vhid = resp["_id"]


@step("Cannot rename to existing collection")
@async_run_until_complete
async def step_impl(context: Context):
    assert_that(hasattr(context, "client"), "Client not set")
    context.oid = ObjectId()
    req = rename_collection_request(database=_database, collection=_collection, document=Dict({"target": "test"}))
    resp = await context.client.execute(req)
    _log.info(f"Rename collection response: {dumps(resp)}")

    assert_that(_has_key("error", resp), "Response did not return error")


@step("Rename collection")
@async_run_until_complete
async def step_impl(context: Context):
    assert_that(hasattr(context, "client"), "Client not set")
    context.oid = ObjectId()
    req = rename_collection_request(database=_database, collection=_collection, document=Dict({"target": _target}))
    resp = await context.client.execute(req)
    _log.info(f"Rename collection response: {dumps(resp)}")

    assert_that(_has_key("error", resp), not_(True), "Response contained error")
    assert_that(_has_key("database", resp), "Response did not contain input database")
    assert_that(_has_key("collection", resp), "Response did not contain target collection")


@step("Retrieve document from renamed collection")
@async_run_until_complete
async def step_impl(context: Context):
    assert_that(hasattr(context, "client"), "Client not set")
    assert_that(hasattr(context, "oid"), "Document id not set")
    _log.info(f"Retrieve document by id: {context.oid}")
    await sleep(0.5)
    req = retrieve_request(database=_database, collection=_target, document=Dict({"_id": context.oid}))
    resp = await context.client.execute(req)
    _log.info(f"Retrieve response: {dumps(resp)}")

    """
    assert_that(_has_key("error", resp), not_(True), "Response contained error")
    assert_that(_has_key("results", resp), not_(True), "Response contained results array")
    assert_that(_has_key("result", resp), "Response did not include result document")
    assert_that(_has_key("_id", resp.result), "Response result does not have id")
    assert_that(resp.result["_id"], equal_to(context.oid), "Response result id not same")
    assert_that(_has_key("key", resp.result), "Response result did not contain key")
    assert_that(resp.result.key, equal_to("value"), "Response result key value not same")
    """


@step("Retrieve version history for renamed collection")
@async_run_until_complete
async def step_impl(context: Context):
    assert_that(hasattr(context, "client"), "Client not set")
    assert_that(hasattr(context, "vhdb"), "Version history database not set")
    assert_that(hasattr(context, "vhc"), "Version history collection not set")
    assert_that(hasattr(context, "vhid"), "Version history id not set")
    _log.info(f"Retrieve version history by id: {context.vhid}")
    await sleep(0.5)
    req = retrieve_request(database=context.vhdb, collection=context.vhc,
                           document=Dict({"_id": context.vhid, "database": _database, "collection": _target}))
    resp = await context.client.execute(req)
    _log.info(f"Retrieve version history response: {dumps(resp)}")

    """
    assert_that(_has_key("error", resp), not_(True), "Response contained error")
    assert_that(_has_key("results", resp), not_(True), "Response contained results array")
    assert_that(_has_key("result", resp), "Response did not include result document")
    assert_that(_has_key("_id", resp.result), "Response result did not include id")
    assert_that(resp.result["_id"], equal_to(context.vhid), "Response result id not same")
    assert_that(_has_key("entity", resp.result), "Response result did not include entity")
    assert_that(_has_key("_id", resp.result.entity), "Response result entity did not include id")
    assert_that(resp.result.entity["_id"], equal_to(context.oid), "Response result entity id not same")
    """


@step("Drop renamed collection")
@async_run_until_complete
async def step_impl(context: Context):
    assert_that(hasattr(context, "client"), "Client not set")
    assert_that(hasattr(context, "oid"), "Document id not set")
    req = drop_collection_request(database=_database, collection=_target, document=Dict({"clearVersionHistory": True}))
    resp = await context.client.execute(req)
    _log.info(f"Drop collection response: {dumps(resp)}")

    assert_that(_has_key("error", resp), not_(True), "Response contained error")
    assert_that(_has_key("results", resp), not_(True), "Response contained results array")
    assert_that(_has_key("result", resp), not_(True), "Response contained result document")
    assert_that(_has_key("dropCollection", resp), "Response did not include dropCollection")


@step("Revision history cleared for dropped collection")
@async_run_until_complete
async def step_impl(context: Context):
    assert_that(hasattr(context, "client"), "Client not set")
    assert_that(hasattr(context, "vhdb"), "Version history database not set")
    assert_that(hasattr(context, "vhc"), "Version history collection not set")
    assert_that(hasattr(context, "vhid"), "Version history id not set")
    req = retrieve_request(database=context.vhdb, collection=context.vhc,
                           document=Dict({"_id": context.vhid, "database": _database, "collection": _target}))
    resp = await context.client.execute(req)
    _log.info(f"Retrieve version history response: {dumps(resp)}")

    assert_that(_has_key("error", resp), "Response not error")

File: client/julia/MongoService/test/rename.jl
import MongoService: Client, CreateRequest, DropCollectionRequest, RenameCollectionRequest, RetrieveRequest, Document, execute
import LightBSON: BSONObjectId

@testset "Rename collection test suite" begin
    c = Client(2000, "julia-test")
    id = BSONObjectId()
    db = "itest"
    col = "oldname"
    target = "newname"
    count = 0

    vhdb = ""
    vhc = ""
    vhoid = BSONObjectId()

    @testset "Document created in a new collection" begin
        doc = Document("_id" => id, "key" => "value")
        req = CreateRequest(database=db, collection=col, document=doc)
        resp = execute(c, req)
        @info "Rename: create: $resp"

        @test haskey(resp, "error") == false
        @test haskey(resp, "database")
        @test haskey(resp, "collection")
        @test haskey(resp, "_id")
        @test haskey(resp, "entity")
        @test resp["entity"] == id

        vhdb = resp["database"]
        vhc = resp["collection"]
        vhoid = resp["_id"]
    end

    @testset "Cannot rename to existing collection" begin
        doc = Document("target" => "test")
        req = RenameCollectionRequest(database=db, collection=col, document=doc)
        resp = execute(c, req)
        @info "Rename: existing: $resp"

        @test haskey(resp, "error")
    end

    @testset "Rename collection" begin
        doc = Document("target" => target)
        req = RenameCollectionRequest(database=db, collection=col, document=doc)
        resp = execute(c, req)
        @info "Rename: rename: $resp"

        @test haskey(resp, "error") == false
        @test haskey(resp, "database")
        @test haskey(resp, "collection")
        @test resp["database"] == db
        @test resp["collection"] == target
    end

    @testset "Retrieve document from renamed collection" begin
        doc = Document("_id" => id)
        req = RetrieveRequest(database=db, collection=target, document=doc)
        resp = execute(c, req)
        @info "Rename: retrieve by id: $resp"

        @test haskey(resp, "error") == false
        @test haskey(resp, "result")
        @test haskey(resp, "results") == false
        @test haskey(resp["result"], "_id")
        @test resp["result"]["_id"] == id
        @test haskey(resp["result"], "key")
        @test resp["result"]["key"] == "value"
    end

    @testset "Retrieve version history for renamed collection" begin
        doc = Document("_id" => vhoid, "database" => db, "collection" => target)
        req = RetrieveRequest(database=vhdb, collection=vhc, document=doc)
        resp = execute(c, req)
        @info "Rename: retrieve version history by id: $resp"

        @test haskey(resp, "error") == false
        @test haskey(resp, "result")
        @test haskey(resp, "results") == false
        @test haskey(resp["result"], "_id")
        @test resp["result"]["_id"] == vhoid
        @test haskey(resp["result"], "entity")
        @test haskey(resp["result"]["entity"], "_id")
        @test resp["result"]["entity"]["_id"] == id
    end

    @testset "Drop renamed collection" begin
        doc = Document("clearVersionHistory" => true)
        req = DropCollectionRequest(database=db, collection=target, document=doc)
        resp = execute(c, req)
        @info "Rename: drop collection: $resp"

        @test haskey(resp, "error") == false
        @test haskey(resp, "dropCollection")
    end

    @testset "Revision history cleared for dropped collection" begin
        sleep(0.25)
        doc = Document("_id" => vhoid, "database" => db, "collection" => target)
        req = RetrieveRequest(database=vhdb, collection=vhc, document=doc)
        resp = execute(c, req)
        @info "Rename: retrieve version history by id after clear: $resp"

        @test haskey(resp, "error")
    end
end
